tornado/tcpclient.py

```diff
@@ -173,19 +174,30 @@ def connect(self, host, port, af=socket.AF_UNSPEC, ssl_options=None,
         .. versionchanged:: 4.5
            Added the ``source_ip`` and ``source_port`` arguments.
         """
-        addrinfo = yield self.resolver.resolve(host, port, af)
+        if timeout:
+            addrinfo = yield gen.with_timeout(
+                timeout, self.resolver.resolve(host, port, af))
+        else:
+            addrinfo = yield self.resolver.resolve(host, port, af)
         connector = _Connector(
             addrinfo, self.io_loop,
             functools.partial(self._create_stream, max_buffer_size,
                               source_ip=source_ip, source_port=source_port)
         )
-        af, addr, stream = yield connector.start()
+        if timeout:
+            af, addr, stream = yield gen.with_timeout(timeout, connector.start())
```

Comment:
I don't think this solves the whole problem. I think the timeout needs to be passed down into the connector too. We don't yet have a way to cancel an ongoing connection attempt, but we can at least stop creating new ones once the deadline is reached.